drivers/wifi/esp/esp_offload.c

```diff
@@ -0,0 +1,706 @@
+			      addr_str, sizeof(addr_str));
+		snprintk(cmd_buf, sizeof(cmd_buf),
+			 "AT+CIPSEND=%d,%d,\"%s\",%d",
+			 sock->link_id, pkt_len, addr_str,
+			 ntohs(net_sin(&sock->dst)->sin_port));
+	}
+
+	k_sem_take(&dev->cmd_handler_data.sem_tx_lock, K_FOREVER);
+	k_sem_reset(&dev->sem_tx_ready);
+
+	ret = modem_cmd_send_nolock(&dev->mctx.iface, &dev->mctx.cmd_handler,
+			     NULL, 0, cmd_buf, &dev->sem_response,
+			     ESP_CMD_TIMEOUT);
+	if (ret < 0) {
+		LOG_DBG("Failed to send command");
+		goto out;
+	}
+
+	ret = modem_cmd_handler_update_cmds(&dev->cmd_handler_data,
+					    cmds, ARRAY_SIZE(cmds),
+					    true);
+	if (ret < 0) {
+		goto out;
+	}
+
+	/*
+	 * After modem handlers have been updated the receive buffer
+	 * needs to be processed again since there might now be a match.
+	 */
+	k_sem_give(&dev->iface_data.rx_sem);
+
+	/* Wait for '>' */
+	ret = k_sem_take(&dev->sem_tx_ready, 5000);
+	if (ret < 0) {
+		LOG_DBG("Timeout waiting for tx");
+		goto out;
+	}
+
+	while (pkt_len) {
+		write_len = MIN(pkt_len, sizeof(dev->pkt_buf));
+		net_pkt_read(sock->tx_pkt, dev->pkt_buf, write_len);
+		dev->mctx.iface.write(&dev->mctx.iface, dev->pkt_buf,
+				      write_len);
+		pkt_len -= write_len;
+	}
+
+	/* Wait for 'SEND OK' or 'SEND FAIL' */
+	k_sem_reset(&dev->sem_response);
+	ret = k_sem_take(&dev->sem_response, ESP_CMD_TIMEOUT);
+	if (ret < 0) {
+		LOG_DBG("No send response");
+		goto out;
+	}
+
+	ret = modem_cmd_handler_get_error(&dev->cmd_handler_data);
+	if (ret != 0) {
+		LOG_DBG("Failed to send data");
+	}
+
+out:
+	(void)modem_cmd_handler_update_cmds(&dev->cmd_handler_data,
+					    NULL, 0U, false);
+	k_sem_give(&dev->cmd_handler_data.sem_tx_lock);
+
+	net_pkt_unref(sock->tx_pkt);
+	sock->tx_pkt = NULL;
+
+	return ret;
+}
+
+static void esp_send_work(struct k_work *work)
+{
+	struct esp_socket *sock;
+	struct esp_data *dev;
+	int ret = 0;
+
+	sock = CONTAINER_OF(work, struct esp_socket, send_work);
+	dev = esp_socket_to_dev(sock);
+
+	if (!esp_socket_in_use(sock)) {
+		LOG_DBG("Socket %d not in use", sock->idx);
+		return;
+	}
+
+	ret = _sock_send(dev, sock);
+	if (ret < 0) {
+		LOG_ERR("Failed to send data: link %d, ret %d", sock->link_id,
+			ret);
+	}
+
+	if (sock->send_cb) {
+		sock->send_cb(sock->context, ret, sock->send_user_data);
+	}
+}
+
+static int esp_sendto(struct net_pkt *pkt,
+		      const struct sockaddr *dst_addr,
+		      socklen_t addrlen,
+		      net_context_send_cb_t cb,
+		      s32_t timeout,
+		      void *user_data)
+{
+	struct net_context *context;
+	struct esp_socket *sock;
+	struct esp_data *dev;
+	int ret = 0;
+
+	context = pkt->context;
+	sock = (struct esp_socket *)context->offload_context;
+	dev = esp_socket_to_dev(sock);
+
+	LOG_DBG("link %d, timeout %d", sock->link_id, timeout);
+
+	if (sock->tx_pkt) {
+		return -EBUSY;
+	}
+
+	if (sock->type == SOCK_STREAM) {
+		if (!esp_socket_connected(sock)) {
+			return -ENOTCONN;
+		}
+	} else {
+		if (!esp_socket_connected(sock)) {
```

Comment:
Does esp require that one can only use connected UDP sockets?

Reply:
What it does require is that you have issued a `CIPSTART` command to allocate a linkid. That command requires and initial remote ip and port and so it sort of becomes connected.

I think changing endpoint might be supported in new firmwares but I have not tested it and so it is not supported currently in the driver.